File: p1/gth773s/java/src/abcd/Diff.java
package abcd;

class Diff {

    private Value a;
    private Value b;

    Diff(Value a, Value b) {
        this.a = a;
        this.b = b;
    }

    Value a() {
        return a;
    }

    Value b() {
        return b;
    }

    boolean isConstant() {
        return a == b || (a.isConstant() && b.isConstant());
    }

    Integer constantValue() {
        if (a == b) return 0;
        if (a.isConstant() && b.isConstant()) return b.constant - a.constant;
        return null;
    }

    LBound fixedBound() {
        Integer value = constantValue();
        return value == null ? null : LBound.fixed(value);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Diff diff = (Diff) o;
        if (a != null ? !a.equals(diff.a) : diff.a != null) return false;
        if (b != null ? !b.equals(diff.b) : diff.b != null) return false;
        return true;
    }

    @Override
    public int hashCode() {
        int result = a != null ? a.hashCode() : 0;
        result = 31 * result + (b != null ? b.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return "(" + a + ", " + b + ")";
    }

}

File: p1/gth773s/phase1-gth773s/Diff.cpp
#include "PhaseOne.h"

using namespace llvm;
using namespace std;

namespace phaseone {

  Diff Diff::create(Value *a, Value *b) {
    Diff diff;
    diff._a = a;
    diff._b = b;
    return diff;
  }

  bool Diff::isConstant() {
    return (_a == _b) || (isa<ConstantInt>(_a) && isa<ConstantInt>(_b));
  }

  Integer Diff::constantValue() {
    if (_a == _b) return Integer::some(APInt(1, 0));
    if (ConstantInt *cA = dyn_cast<ConstantInt>(_a)) {
      if (ConstantInt *cB = dyn_cast<ConstantInt>(_b)) {
        return Integer::some(cB->getValue() - cA->getValue());
      }
    }
    return Integer::none();
  }

  L Diff::fixedBoundGet(APInt c) {
    Integer cv = constantValue();
    if (!cv.hasValue()) return L_NULL;
    LBound *bound = LBound::fixed(cv.value());
    L l = bound->get(c);
    delete bound;
    return l;
  }

  bool Diff::operator<(const Diff &o) const {
    return (_a != o._a) ? (_a < o._a) : (_b < o._b);
  }

}
